File: libraries/content/src/main/kotlin/orwir/starrit/content/feed/Feed.kt
package orwir.starrit.content.feed

import androidx.lifecycle.LiveData
import androidx.paging.PagedList
import orwir.starrit.content.post.Post
import orwir.starrit.core.model.ActionHolder
import orwir.starrit.core.model.NetworkState
import orwir.starrit.core.util.createHashCode
import java.io.Serializable
import java.util.*

internal const val HOME = "/r/home"
internal const val POPULAR = "/r/popular"
internal const val ALL = "/r/all"

class Feed internal constructor(
    val posts: LiveData<PagedList<Post>>,
    val networkState: LiveData<NetworkState>,
    val retry: ActionHolder
) {

    sealed class Type(val path: String, val composite: Boolean = true) : Serializable {
        object Home : Type(HOME)
        object Popular : Type(POPULAR)
        object All : Type(ALL)
        class Subreddit(name: String) : Type("/r/$name", false) {
            override fun equals(other: Any?) = other is Subreddit && path == other.path
            override fun hashCode() = createHashCode(path)
        }

        override fun toString(): String = this::class.simpleName ?: path

        internal fun asParameter() = if (this is Home) "" else path
    }

    enum class Sort {
        Best,
        Hot,
        New,
        Top,
        Rising;

        internal fun asParameter() = name.toLowerCase(Locale.ENGLISH)
    }

}

File: lib/models/feed.dart
import 'package:flutter/foundation.dart';
import 'package:starrit/extensions/object.dart';

@immutable
class Feed {
  const Feed.home({Sort sort = Sort.best}) : this._(Type.home, sort);
  const Feed.popular({Sort sort = Sort.best}) : this._(Type.popular, sort);
  const Feed.all({Sort sort = Sort.best}) : this._(Type.all, sort);
  Feed.subreddit(String subreddit, {Sort sort = Sort.best})
      : this._(
          Type.subreddit(subreddit),
          sort,
        );
  const Feed._(this.type, this.sort);

  final Type type;
  final Sort sort;

  String get label => '${type.label}${sort.label}';

  @override
  String toString() => '$type$sort';

  @override
  int get hashCode => hash([type, sort]);

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Feed &&
          runtimeType == other.runtimeType &&
          type == other.type &&
          sort == other.sort;
}

@immutable
class Type {
  static const home = Type._('');
  static const popular = Type._('/r/popular');
  static const all = Type._('/r/all');

  static const Iterable<Type> values = [
    Type.home,
    Type.popular,
    Type.all,
  ];

  const Type.subreddit(this.path);
  const Type._(this.path);

  final String path;

  String get label => this == Type.home ? '/home' : toString();

  @override
  String toString() => path;

  @override
  int get hashCode => path.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Type && runtimeType == other.runtimeType && path == other.path;
}

@immutable
class Sort {
  static const best = Sort._('/best');
  static const hot = Sort._('/hot');
  static const newest = Sort._('/new');
  static const top = Sort._('/top');
  static const rising = Sort._('/rising');
  static const controversial = Sort._('/controversial');

  static const Iterable<Sort> values = [
    Sort.best,
    Sort.hot,
    Sort.newest,
    Sort.top,
    Sort.rising,
    Sort.controversial,
  ];

  const Sort._(this.path);

  final path;

  String get label => toString();

  @override
  String toString() => path;

  @override
  int get hashCode => path.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Type && runtimeType == other.runtimeType && path == other.path;
}
